### PR TITLE
Updating docs and synching protocols with latest generator

### PR DESCRIPTION
**Pull request by 5A11**
## Proposed changes

Update generator documentation
Updated all existing protocols **except default** with the output of the latest protocol generator.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/protocol-generator.md [hunk 1]
```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
...
+| Primitive mapping types<sup>2</sup>    | `<PMT>` | `pt:dict[<PT>, <PT>]`                                                                 | `pt:dict[pt:str, pt:bool]`               | `Dict[str, bool]`                  |
+| Multi types                | `<MT>`  | `pt:union[<PT>/<CT>/<PCT>/<PMT>, ..., <PT>/<CT>/<PCT>/<PMT>]` | `pt:union[ct:DataModel, pt:set[pt:str]]` | `Union[DataModel, FrozenSet[str]]` |
+| Optional types             | `<O>`   | `pt:optional[<MT>/<PMT>/<PCT>/<PT>/<CT>]`                                       | `pt:optional[pt:bool]`                   | `Optional[bool]`                   |
 
 <sup>*</sup> <sub>This is how variable length tuples containing elements of the same type are declared in Python; see <a href="https://docs.python.org/3/library/typing.html#typing.Tuple" target=_blank>here</a> </sub>
 
+#### Notes
+
+1. Currently, there is no way to describe custom types in a language independent format. This means that if a protocol specification includes custom types, the required implementations must be provided manally.
```
Comment (DavidMinarsch) [inline docs/protocol-generator.md:111]: `there is no way to describe custom types in a target language independent format` ..or something like that, language independent does not make sense as the specification language is also a language..

File: docs/protocol-generator.md [hunk 2]
```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
...
+    * Before generating the protocol, the protocol buffer schema code for every custom type must be provided in the protocol specification. You can see an example of this in the above protocol specification. In particular, if a specification includes custom types, the specification must include a second yaml document (each yaml document is enclosed within `---` and `...`) which for every custom type describes its protocol buffer schema code.   
+    * Once the generator is called, it produces a `custom_types` module containing stub implmentations for every custom type in the specification. The user must then modify this module and add implemntations for every custom type in the specification. This includes implementations of how an object of a custom type can be encoded and decoded using protocol buffer.
```
Comment (DavidMinarsch) [inline docs/protocol-generator.md:113]: add implemntations > add implementations

File: docs/protocol-generator.md [hunk 3]
```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
...
+    * Note, currently the way custom types are dealt with in the generator is admittedly inconvinient. The reason is, the generator does not know the structure of custom types and how they may be serialised/deserialised. Although this approach works, it is only a temporary solution until further work on an infrastructure-independent type description language is finished (similar to how the generator is designed to be an infrastructure-independent protocol description language).
```
Comment (DavidMinarsch) [inline docs/protocol-generator.md:114]: I would not call it infrastructure-independent..

File: docs/protocol-generator.md [hunk 4]
```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
...
+2. Currently, the first element in `pt:dict` cannot be a `<CT>`, `pt:float` or  `pt:bytes`. This is because of a constraint in protocol buffer version 3 which is the framework's underlying serialisation mechanism. In a future version, we may address this limitation, in which case we will relax this constraint.
+3. In protocol buffer version 3, which is the version used by the generator, there is no way to check whether an optional field (i.e. contents of type `pt:optional[...]`) has been set or not (see discussion <a href="https://github.com/protocolbuffers/protobuf/issues/1606" target=_blank>here</a>). In proto3, all optional fields are assigned a default value (e.g. `0` for integers types, `false` for boolean types, etc). Therefore, there is no way to know whether an optional field whose value is the default value is not set, or is set but its value happens to be the default value. Because of this, in the generated protocol schema file (the `.proto` file), for every optional content there is a second field that declares whether this field is set or not. We will maintain this temporary solution until a cleaner alternative is found.
```
Comment (DavidMinarsch) [inline docs/protocol-generator.md:116]: there is no way to know whether an optional field whose value is the default value is not set > there is no way to know from an optional field itself whether an optional field whose value is the default value is not set

File: docs/protocol-generator.md [hunk 5]
```diff
@@ -30,40 +30,50 @@ As such, it needs to follow the <a href="https://pyyaml.org/wiki/PyYAMLDocumenta
...
 
 
 Each protocol specification yaml file contains some basic information about the protocol. The allowed fields and what they represent are:
 
  * `name`: The name of the protocol (written in <a href="https://en.wikipedia.org/wiki/Snake_case" target="_blank">snake_case</a>)
- * `authors`: List of authors
+ * `author`: The creator of the protocol
  * `version`: The current version of the protocol
  * `license`: Licensing information
+ * `aea_version`: The version(s) of the framework that support this protocol
```
Comment (DavidMinarsch) [inline docs/protocol-generator.md:63]: must follow: https://www.python.org/dev/peps/pep-0440/

File: docs/protocol-generator.md (hunk 1 above)
Comment (5A11) [reply docs/protocol-generator.md:111]: Done

File: docs/protocol-generator.md (hunk 2 above)
Comment (5A11) [reply docs/protocol-generator.md:113]: Done

File: docs/protocol-generator.md (hunk 3 above)
Comment (5A11) [reply docs/protocol-generator.md:114]: Done

File: docs/protocol-generator.md (hunk 4 above)
Comment (5A11) [reply docs/protocol-generator.md:116]: Done

File: docs/protocol-generator.md (hunk 5 above)
Comment (5A11) [reply docs/protocol-generator.md:63]: Done

**Pull request merged**